File: Code/Ch1-memory-centric-system-model/advanced-model/abstract_system_code.py
from abstract_system_constants import *
from abstract_system_encoder import *

FIB_MAX=84
def load_code(ramState,cfg):
    (PERIPHERALS,HL,hlCode)=cfg
    if HL:
        def fib_hl( systemState,registers ):
            (r1,r2,r4)=(1,1,FIB_MAX)
            while r4!=0:
                r3=r1+r2
                r1=r2
                r2=r3
                r4-=1
                systemState[r4]=r3
            registers[1:5]=[r1,r2,r3,r4]        
            return ( systemState,registers )  
#        hlCode[0]=fib_hl
        fib_prog=[[HLI,fib_hl]]
#        ramState[CODE]=0
    else:
        # _=0
        # High-level instructions go here
        def fib_hl( systemState,registers ):
            # (r1,r2,r4)=(1,1,FIB_MAX)
            r1 = int(registers[1]) 
            r2 = int(registers[2])
            r3 = int(registers[2])
            r4 = int(registers[4])
            while r4!=0:
                r3=r1+r2
                r1=r2
                r2=r3
                r4-=1
                systemState[r4]=r3
            registers[1:5]=[r1,r2,r3,r4]
            return ( systemState,registers )
        def timerIRS(systemState,registers):
            print( 'IRS Called for Timer')
            return ( systemState,registers )
        loop = 1 # declare a label, the value is irrelevant as long as it's unique
        print(TIMER)
        fib_prog=[
        [SET,R1,10],
        [SET,R2,10], # start periodic timer
        [SET,R3,TIMER],
        [STR,R1,R3],
        [SET,R3,TIMER+1],
        [STR,R2,R3],
        [SET,R1,1],
        [SET,R2,1],
        [SET,R3,0],
        [SET,R4,FIB_MAX],
        # # [HLI,fib_hl], # example of how to mix high-level instructions with low-level ones
        [SET,R5,1],
        (loop,[ADD,R3,R1,R2]),
        [MOV,R1,R2],
        [MOV,R2,R3],
        [SUB,R4,R4,R5],
        [STR,R3,R4],
        [CBNZ,R4,loop],
        [WFI],
        [HLI, timerIRS ]
        ]
        
    #print(fib_prog)
    fib_iws=encodeProgram(fib_prog)
#        pc=0
#        for iw in fib_iws:
#            decoded_instr = decodeInstruction(iw)
#            print(pc,decoded_instr)
#            pc+=1
    pc=0
    for iw in fib_iws:
        if type(iw)==tuple:
            hliw,hli=iw
            # print("PC:",pc)
            hlCode[pc] = hli
            ramState[CODE+pc] = hliw
        else:
            ramState[CODE+pc] = iw
        pc+=1
        
    cfg=(PERIPHERALS,HL,hlCode)            
    return (ramState,cfg)

File: Code/Ch1-memory-centric-system-model/cpp-code/abstract_system_code.cc
#include <tuple>
#ifdef VERBOSE
#include <iostream>
#endif
#include "abstract_system_constants.h"
#include "abstract_system_decoder.h"
#include "abstract_system_encoder.h"
#include "abstract_system_code.h"

std::vector<uint64_t>& load_code(std::vector<uint64_t>& ramState){
    uint64_t _ = 0;
    uint64_t loop = 1;

    std::vector< std::vector<uint64_t>> fib_prog={
        {_,SET,R1,1,_},
        {_,SET,R2,1,_},
        {_,SET,R3,0,_},
        {_,SET,R4,FIB_MAX,_},
        {_,SET,R5,1,_},
        {loop,ADD,R3,R1,R2},
        {_,MOV,R1,R2,_},
        {_,MOV,R2,R3,_},
        {_,SUB,R4,R4,R5},
        {_,STR,R3,R4,_},
        {_,CBNZ,R4,loop,_},
        {_,WFI,_,_,_}
    };
    std::vector<uint64_t>   fib_iws=encodeProgram(fib_prog);
    uint64_t pc=0;
    for (auto iw : fib_iws) {
#ifdef VERBOSE        
        std::cout << pc<<" : "<<iw << "\n";
#endif
        std::tuple<uint64_t,std::tuple<uint64_t,uint64_t,uint64_t>> decoded_instr=decodeInstruction(iw);
        uint64_t v1;
        std::tuple<uint64_t,uint64_t,uint64_t> t2;
        uint64_t v2,v3,v4;
        std::tie(v1,t2)=decoded_instr;
        std::tie(v2,v3,v4)=t2;
#ifdef VERBOSE
        std::cout << pc<<" : " <<  v1 << ";" << v2 << "," <<  v3<< "," <<  v4 << "\n";
#endif
        pc+=1;
    }
#ifdef VERBOSE
    std::cout << "CODE: "<<CODE<<"\n";
#endif
    pc=0;
    for (uint64_t iw : fib_iws ) {
#ifdef VERBOSE
        std::cout << "write IW " << iw << " into " << CODE+pc << "\n";
#endif
        ramState[CODE+pc]=iw;
        pc+=1;
    }
    return ramState;
}
